zrml/neo-swaps/src/types/liquidity_tree.rs

```diff
@@ -0,0 +1,1205 @@
+                        lazy_fees: Zero::zero(),
+                    },
+                    Node::<Runtime> {
+                        account: None,
+                        stake: Zero::zero(),
+                        fees: Zero::zero(),
+                        descendant_stake: Zero::zero(),
+                        lazy_fees: Zero::zero(),
+                    },
+                    // Depth 3
+                    Node::<Runtime> {
+                        account: Some(6),
+                        stake: _12,
+                        fees: _1,
+                        descendant_stake: Zero::zero(),
+                        lazy_fees: _3,
+                    },
+                    Node::<Runtime> {
+                        account: None,
+                        stake: Zero::zero(),
+                        fees: Zero::zero(),
+                        descendant_stake: Zero::zero(),
+                        lazy_fees: Zero::zero(),
+                    },
+                    Node::<Runtime> {
+                        account: Some(8),
+                        stake: _4,
+                        fees: _1,
+                        descendant_stake: Zero::zero(),
+                        lazy_fees: Zero::zero(),
+                    },
+                ]
+                .try_into()
+                .unwrap(),
+                account_to_index:
+                    create_b_tree_map!({3 => 0, 9 => 2, 5 => 3, 7 => 4, 6 => 7, 8 => 9})
+                        .try_into()
+                        .unwrap(),
+                abandoned_nodes: vec![1, 5, 6, 8].try_into().unwrap(),
+            }
+        }
+
+        /// Create a full tree. All nodes have the same stake of 1.
+        pub(super) fn create_full_tree() -> LiquidityTreeOf<Runtime> {
+            let max_depth = LiquidityTreeOf::<Runtime>::max_depth();
+            let node_count = LiquidityTreeOf::<Runtime>::max_node_count();
+            let nodes = (0..node_count)
+                .map(|a| Node::<Runtime>::new(a as u128, 1))
+                .collect::<Vec<_>>()
+                .try_into()
+                .unwrap();
+            let account_to_index = (0..node_count)
+                .map(|a| (a as u128, a))
+                .collect::<BTreeMap<_, _>>()
+                .try_into()
+                .unwrap();
+            let mut tree = LiquidityTreeOf::<Runtime> {
+                nodes,
+                account_to_index,
+                abandoned_nodes: vec![].try_into().unwrap(),
+            };
+            // Nodes have the wrong descendant stake at this point, so let's fix that.
+            for (index, node) in tree.nodes.iter_mut().enumerate() {
+                let exp = max_depth + 1 - log2(index as u32 + 1);
+                node.descendant_stake = 2u128.pow(exp) - 2;
+            }
+            tree
+        }
+
+        fn log2(x: u32) -> u32 {
+            std::mem::size_of::<u32>() as u32 * 8 - x.leading_zeros() - 1
+        }
```

Comment:
The log is pretty standard using the fact that the standard binary representation is built on powers of two, but I've noticed that there's also a builtin: 1f30f4691f09e1394ea09e1597295d1fb2abadcf

Reply:
To understand the descendant stake, since all nodes have stake `1`, the question is: How many descendants does a node in a binary tree of a given size have as a function of its index. This is best done by just counting (unless you want an actual proof). Obviously, the result should be exponential and the exponent should be an inverse of the depth.

Again, best to just draw an example.

Reply:
https://github.com/zeitgeistpm/zeitgeist/commit/1f30f4691f09e1394ea09e1597295d1fb2abadcf usage of `unwrap`